Queue.cpp

```diff
@@ -65,29 +74,50 @@ class Queue::Impl
         }
     }
 
-    std::unique_ptr<Msg> request(Msg&& msg)
+    std::unique_ptr<Msg> request(Msg&& msg, int timeoutMillis)
     {
-        // Construct an ad hoc Queue to handle response Msg
+        Request req;
+
+        // emplace the request in the map
         std::unique_lock<std::mutex> lock(responseMapMutex_);
         auto it = responseMap_.emplace(
-            std::make_pair(msg.getUniqueId(), std::unique_ptr<Queue>(new Queue))).first;
-        lock.unlock();
+            std::make_pair(msg.getUniqueId(), &req)).first;
 
         put(std::move(msg));
-        auto response = it->second->get(); // Block until response is put to the response Queue
 
-        lock.lock();
-        responseMap_.erase(it); // Delete the response Queue
-        lock.unlock();
+        // wait_for returns false if the return is due to timeout
+        if (timeoutMillis <= 0)
+            req.condVar.wait(lock, [&req]{return req.response.get();});
+        else
+        {
+            auto timeoutOccured = !req.condVar.wait_for(
+            lock,
+            std::chrono::milliseconds(timeoutMillis),
+            [&req] {return req.response.get();});
+
+            if (timeoutOccured)
+            {
+                // remove the request from the map, as it is on the stack
```

Comment:
I don't really get this comment: the Request object is not needed anymore so it shouldn't matter if it's in callers stack?